### ISSUE TITLE
Basis for categorical variables

### ISSUE THREAD

**Issue by BalzaniEdoardo** [flatironinstitute/nemos#486]
**Is your feature request related to a problem? Please describe.**
Add basis that simplifies processing of categorical variables and interaction between categories and other covaraites. This seems to be a common design matrix construction pattern, see discussion #473 . 

**Describe the solution you'd like**
A simple solution would be a basis that maps a vector of categories to a 1-hot encoded output. This type of basis, multiplied with any other basis, will generate an interaction `category x predictor`, allowing learning one predictor per category. It could also be used directly to form a predictor that learns an offset per category.

```python
>>> import numpy as np
>>> import nemos as nmo

>>> # Example API
>>> trial_type_basis = nmo.basis.Category(n_categories=3)
>>> trial_types = np.array(["A", "A", "B", "C"])

>>> X = trial_type_basis.compute_features(trial_types)
>>> X
Array([[1., 0., 0.],
       [1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]], dtype=float32)

>>> # For an interaction with a covariate
>>> position = np.arange(x.shape[0])
>>> pos_basis = nmo.basis.RaisedCosineLinearEval(5)
>>> interaction = pos_basis * trial_type_basis
>>> X = interaction.compute_features(position, trial_types)
Array([[1.        , 0.        , 0.        , 0.49999997, 0.        ,
        0.        , 0.        , 0.        , 0.        , 0.        ,
        0.        , 0.        , 0.        , 0.        , 0.        ],
       [0.24999997, 0.        , 0.        , 0.9330127 , 0.        ,
        0.        , 0.75      , 0.        , 0.        , 0.06698731,
        0.        , 0.        , 0.        , 0.        , 0.        ],
       [0.        , 0.        , 0.        , 0.        , 0.06698725,
        0.        , 0.        , 0.74999994, 0.        , 0.        ,
        0.9330127 , 0.        , 0.        , 0.25000006, 0.        ],
       [0.        , 0.        , 0.        , 0.        , 0.        ,
        0.        , 0.        , 0.        , 0.        , 0.        ,
        0.        , 0.49999997, 0.        , 0.        , 1.        ]],      dtype=float32)

```

Currently, a way to get to this behavior is to define a custom basis:

```python
>>> n_categories = 3
>>> def func(x):
...     return jax.nn.one_hot(x, n_categories)
>>>  trial_type_basis = nmo.basis.CustomBasis(func)

```

However, handling label encoding/decoding would be up to the user, and requires substantial of boiler plait code to define the custom basis.

**Describe alternatives you've considered**
In #473, we discussed having a dedicated basis that computes the interaction.

```python
import jax.numpy as jnp
import jax
import nemos as nmo

# Receives as input the categories and the interaction
interaction = nmo.basis.CategoryInteraction(n_categories, basis)

X1 = interaction.compute_features(categories, x)
# internally what this does is
X2 = jnpeinsum(
    "tk, tc -> tck", 
    basis.compute_features(x),
    jax.nn.one_hot(categories, n_categories)
)
assert jnp.all(X1 == X2)
```

I think this is a common use case, but it is equivalent to having a basis handling the categories 1-hot encoding and using multiplication to capture interactions. I think this would be not too bad in terms of how much extra code is needed, and also fairly readable.

**Additional context**
At this point, the name basis starts to become a stretch This `basis` are mostly input maps. Transformer is a name that is already taken, so we may want to think about if it would make sense to call this a basis or having different categories of transformations, of which basis are one.
